### ISSUE TITLE
[0.10.0-nightly tsm1] influxd takes ~30 minutes to start up

### ISSUE THREAD

**Issue by dswarbrick** [influxdata/influxdb#5311]
Since starting to use tsm1, I've noticed that the startup time of influxd seems to be proportional to the size of the database. On one system, the DB is about 250 GB, containing around 30 measurements and about 600K series. Each shard is approximately 11 GB, and influxd takes a little over a minute to read each shard. Retention period is 21 days.

Log is attached. Version number in log says 0.9, but it's actually built from git commit 7ccbbecf0325098fc16ecf28b084763532b9b4ec. System has 16 cores, 32 GB RAM, running Debian Jessie amd64, kernel 4.3.3. During startup, the data partition has about 1K read IOPS, but is capable of delivering up to 2K read IOPS, limit enforced by cgroups.

[influxd.txt](https://github.com/influxdata/influxdb/files/81886/influxd.txt)

**Comment by jwilder** [comment]
@dswarbrick Could you show me the output of `ls -lah /srv/influxdb/data/perfdb/default/23/` and `influx_inspect dumptsmdev -all /srv/influxdb/data/perfdb/default/23/000000377-000000005.tsm`?

**Comment by dswarbrick** [comment]
```
# ls -lah /srv/influxdb/data/perfdb/default/23/
total 11G
drwxr-xr-x  2 influxdb influxdb 4.0K Jan  7 10:18 .
drwx------ 25 influxdb influxdb 4.0K Jan  7 23:29 ..
-rw-r--r--  1 influxdb influxdb 2.2G Dec 18 05:55 000000189-000000004.tsm
-rw-r--r--  1 influxdb influxdb 493M Dec 18 05:56 000000189-000000005.tsm
-rw-r--r--  1 influxdb influxdb 2.2G Dec 18 11:44 000000377-000000004.tsm
-rw-r--r--  1 influxdb influxdb 483M Dec 18 11:45 000000377-000000005.tsm
-rw-r--r--  1 influxdb influxdb 2.2G Dec 18 17:34 000000566-000000004.tsm
-rw-r--r--  1 influxdb influxdb 491M Dec 18 17:35 000000566-000000005.tsm
-rw-r--r--  1 influxdb influxdb 2.2G Dec 18 23:24 000000755-000000004.tsm
-rw-r--r--  1 influxdb influxdb 488M Dec 18 23:25 000000755-000000005.tsm
-rw-r--r--  1 influxdb influxdb 478M Dec 21 02:39 000000780-000000003.tsm
```

**Comment by dswarbrick** [comment]
@jwilder Output of influx_inspect is about 46 MB raw (4.8 MB .xz), and contains some sensitive information which I'd prefer not to attach to a public issue. What's the best way to get this information to you?

**Comment by jwilder** [comment]
You can email to jason@influxdb.com if that works.

**Comment by jwilder** [comment]
@dswarbrick Would you be able to test #5318 on your system since you are able to reproduce this slow restart more easily?

**Comment by dswarbrick** [comment]
@jwilder I will test it after the 24h window is up on my current golang 1.4 build, which will be in about 7 hours, (unless the oom-kill strikes before that).

**Comment by jwilder** [comment]
@dswarbrick Great.  Thanks!

**Comment by toddboom** [comment]
@jwilder i'll pull this in on my 100GB datasets and let you know how it looks

**Comment by jwilder** [comment]
@toddboom :+1:

**Comment by toddboom** [comment]
@jwilder this seems to have made a huge difference. startups that previously took 7-10 minutes are now finishing in under a minute.

there may be a slight benefit from compactions that happened since the last time i ran the test (resulting in fewer tsm files), but the approximate size on disk is basically the same.

nice work!

**Comment by toddboom** [comment]
i'll wait for @dswarbrick to comment with his results, but based on my observations, this issue can probably now be closed out by #5318.

**Comment by dswarbrick** [comment]
@jwilder My startup time has reduced from over 30 minutes down to just 10 minutes. The majority of the shards load quite quickly, although there were a couple of outliers. What I mainly notice is that there periods of time during the startup lasting maybe 10s, during which only a single core is maxed out, whilst the others are idle. I am attaching the relevant portion of my influxd.log to show the timings.

The read IOPS were quite a lot lower also (no more than about 70 IOPS), apart from when it read the two or three most recent shards, and you can see in the log that those shards took noticeably longer to load. Read IOPS spiked to about 700 briefly, whilst reading those shards.

[influxd.txt](https://github.com/influxdata/influxdb/files/83704/influxd.txt)

**Comment by jwilder** [comment]
@dswarbrick Seems like it has improved.  Shards are still loaded serially, but that will be changed in a future PR which should also help.  Some of the other slowness is related to loading the in-memory database index.  I'm looking into that as well.

**Comment by dswarbrick** [comment]
@jwilder Great, I look forward to testing it :+1:

**Comment by undera** [comment]
My 0.10 server starts for 3 hours with 9GB database. Should I describe diagnostics here or create a separate issue?

**Comment by toddboom** [comment]
The v0.12.0 release has some fixes from #5372 to improve loading times. Closing this out since that plus the prior fixes should have things in a usable state now.
